### PR TITLE
Error message

### PR DESCRIPTION
**Pull request by sharok**

### CODE REVIEW

File: client/base/form.js [hunk 1]
```diff
@@ -22,13 +22,37 @@ export default class Form extends Component {
     return getValue(keyPath, this.state);
   }
 
+  initState () {
+    return {validationMessage: ''}
+  }
```
Comment (atomiomi) [inline client/base/form.js]: Dont't define `initState` in base components. Children component will override it. In this case you can call `this.forceUpdate()` to rerender view

File: client/base/form.js [hunk 2]
```diff
@@ -22,13 +22,37 @@ export default class Form extends Component {
...
+
   showError (code) {
     let message = error(code);
-    alert(message);
+    this.updateValidationMessage(message);
+  }
+
+  updateValidationMessage (message) {
```
Comment (atomiomi) [inline client/base/form.js]: let's call it `alertMessage`

Comment (sharok) [inline client/base/form.js]: WTF? Nooo, it's the stupid name. I think it's better or `showValidationMessage`.

File: client/base/form.js (hunk 1 above)
Comment (sharok) [inline client/base/form.js]: I have to define its in `signup` and `signin`, right?

Comment (atomiomi) [inline client/base/form.js]: No, just call `this.forceUpdate()` in alertMessage

File: client/base/form.js (hunk 2 above)
Comment (atomiomi) [inline client/base/form.js]: Ok, I like it too

File: client/base/form.js [hunk 3]
```diff
@@ -22,13 +22,37 @@ export default class Form extends Component {
...
+    this.setState({validationMessage: message});
   }
 
   handleAPIError (xhr) {
-    this.showError('unexpected_error');
+    let formattedMessage;
+    let response = JSON.parse(xhr.responseText);
```
Comment (atomiomi) [inline client/base/form.js]: you can get it by `responseJSON` field

File: client/base/form.js (hunk 1 above)
Comment (sharok) [inline client/base/form.js]: Where I have to define `validationMessage` ?

File: client/base/form.js [hunk 4]
```diff
@@ -22,13 +22,37 @@ export default class Form extends Component {
...
+
+    var messages = [];
+    var hasFields = _.has(response, 'error.fields');
+    if (hasFields) {
+      _.forEach(response.error.fields, function (field) {
```
Comment (atomiomi) [inline client/base/form.js:41]: use `for (let key in fields)` it's much faster

File: client/base/form.js [hunk 5]
```diff
@@ -22,13 +22,37 @@ export default class Form extends Component {
...
+        messages.push(field.message);
+      });
+    }
+    else {
+      messages.push(response.error.message);
+    }
+
+    formattedMessage = messages.join('\n');
+
+    this.updateValidationMessage(formattedMessage);
```
Comment (atomiomi) [inline client/base/form.js]: let's join this two lines: `this.updateValidationMessage(messages.join('\n'))`;

File: client/views/signin.js [hunk 6]
```diff
@@ -39,6 +39,9 @@ export default class Signin extends Form {
                 <input valueLink={this.linkState('model.password')} type="password" className="m-control" placeholder={lang.fields.password} required />
               </div>
             </div>
+            { this.state.validationMessage
+              ? <div className="m-message"><span><strong>Error:</strong> {this.state.validationMessage}</span></div>
```
Comment (atomiomi) [inline client/views/signin.js]: Move it to separate component `FormValidation`

File: client/base/form.js (hunk 1 above)
Comment (atomiomi) [inline client/base/form.js]: Nowhere

File: client/views/signin.js (hunk 6 above)
Comment (sharok) [inline client/views/signin.js]: maybe `FormValidattionMessage` or `ErrorMessage`?

Comment (atomiomi) [inline client/views/signin.js]: I like `ErrorMessage`

File: client/base/form.js (hunk 4 above)
Comment (sharok) [inline client/base/form.js:41]: You are wrong master: http://jsperf.com/foreach-vs-for-i-in.

Comment (atomiomi) [inline client/base/form.js:41]: (( my bad

Comment (atomiomi) [inline client/base/form.js:41]: So we need to replace it in all project. @mitro in batman too

Comment (sharok) [inline client/base/form.js:41]: English fix: in the whole project :smile:

Comment (atomiomi) [inline client/base/form.js:41]: My english skills will improve during those projects

Comment (atomiomi) [inline client/base/form.js:41]: But do it in separate PR
